File: sleep.java
import java.lang.*;
import java.io.*;
import java.util.StringTokenizer;

public class sleep
{
   public static void main(String args[]) throws Exception
   {
      FileInputStream   infile;
      BufferedReader    inreader;
      String            line;
      int               numPlans;
      int               i;
      int               minSleepInMinutes;
      int               totalSleepInMinutes;
      int               j;
      StringTokenizer   st;
      int               startHour;
      int               startMinute;
      int               endHour;
      int               endMinute;
      int               startTime;
      int               endTime;
      int               diffSleep;

      // Open the input file into a buffered reader
      infile = new FileInputStream("sleep.in");
      inreader = new BufferedReader(new InputStreamReader(infile));

      // Read the first line and get the number of sleep plans
      line = inreader.readLine();
      numPlans = Integer.parseInt(line.trim());

      // Loop through the sleep plans
      for (i=0; i < numPlans; i++)
      {
         // Read the next line and get the minimum number of hours sleep
         // per week
         line = inreader.readLine();
         minSleepInMinutes = Integer.parseInt(line.trim()) * 60;

         // Initialize the total sleep
         totalSleepInMinutes = 0;

         // Go through the 7 days of the week
         for (j=0; j < 7; j++)
         {
            // Read the next line that has the sleep that night
            line = inreader.readLine();

            // Set-up a tokenizer to process the line
            st = new StringTokenizer(line, ":-");

            // Get start hour and minute
            startHour = Integer.parseInt(st.nextToken().trim());
            startMinute = Integer.parseInt(st.nextToken().trim());

            // Get end hour and minute
            endHour = Integer.parseInt(st.nextToken().trim());
            endMinute = Integer.parseInt(st.nextToken().trim());

            // Compute start time in minutes
            startTime = startHour * 60 + startMinute;

            // Compute end time in minutes
            endTime = endHour * 60 + endMinute;

            // If the end time seems to be earlier than the start time, then 
            // it must be on the next day so add a day's worth of minutes to it
            if (endTime <= startTime)
               endTime = endTime + (26 * 60);

            // Add this sleep shift to the total
            totalSleepInMinutes = totalSleepInMinutes + (endTime - startTime);
         }

         // Print the astronaut number
         System.out.print("Astronaut #" + (i+1) + " ");

         // Check whether we got enough sleep and output accordingly
         if (totalSleepInMinutes < minSleepInMinutes)
         {
            // We need more sleep, uh oh!

            // Figure out how much more we need
            diffSleep = minSleepInMinutes - totalSleepInMinutes;

            // Output how many hours and minutes we need
            System.out.println("needs " + (diffSleep / 60) + " hours and " +
                               (diffSleep % 60) + " minutes more sleep!");
         }
         else
         {
            // We got enough sleep!
            System.out.println("will sleep enough!");
         }
      }
   }
}

File: sleep.cpp
/*
Solution to Problem sleep, UCFHSPT2005

@author yiuyuho@yahoo.com
@date 2005-05-02
*/

#include <fstream>
#include <iostream>

using namespace std;

/*
This function takes a time in HH:MM format and converts it to the number
of minutes from 00:00, so 00:01 will return 1, 12:00 will return 12*60=720
*/
int parse(char *s)
{
	int hr,min;
	//get hours and minute from the characters
	hr=(s[0]-'0')*10+s[1]-'0';
	min=(s[3]-'0')*10+s[4]-'0';
	//every hour has 60 minutes
	return hr*60+min;
}

int main()
{
	//open file
	ifstream cin("sleep.in");
	int N,cc,need,get,i,a,b,day;
	char time[256];
	//number of minutes in a day
	day=26*60;
	//get input size
	cin>>N;
	for(cc=1;cc<=N;cc++)
	{
		//read hours needed
		cin>>need;
		//convert need to minutes, everything will be handled in minutes format.
		need*=60;
		//go through the 7 lines and see how many minutes of sleep is obtained.
		get=0;
		for(i=0;i<7;i++)
		{
			cin>>time;
			a=parse(time);
			b=parse(time+6);
			//if b<=a, then we crossed over 00:00, and b is in fact the next day with reference to a
			//thus we add a day into b
			//keep in mind that you will always sleep, and at most for 26 hours.
			if(b<=a) b+=day;
			//the amount of minutes between a and b is how much we slept
			get+=b-a;
		}
		//output
		cout<<"Astronaut #"<<cc<<" ";
		if(get>=need)
		{
			cout<<"will sleep enough!"<<endl;
		}
		else
		{
			need-=get;
			cout<<"needs "<<need/60<<" hours and "<<need%60<<" minutes more sleep!"<<endl;
		}
	}
	return 0;
}